cursorless.nvim/lua/cursorless/init.lua

```diff
@@ -0,0 +1,131 @@
+-- local Config = require('cursorless.config')
+
+-- We can't use that yet as then when we use load_extensions() we get an error.
+-- So instead atm we rely on registering the functions from vim script
+-- local function register_functions1()
+--   local path = require('cursorless.utils').cursorless_nvim_path()
+--   vim.api.nvim_call_function('remote#host#RegisterPlugin', {
+--     'node',
+--     path .. '/rplugin/node/command-server',
+--     {
+--       { sync = false, name = 'CommandServerLoadExtension', type = 'function', opts = {} },
+--       { sync = false, name = 'CommandServerRunCommand', type = 'function', opts = {} },
+--     },
+--   })
+--   vim.api.nvim_call_function('remote#host#RegisterPlugin', {
+--     'node',
+--     path .. '/rplugin/node/cursorless-neovim',
+--     {
+--       { sync = false, name = 'CursorlessLoadExtension', type = 'function', opts = {} },
+--     },
+--   })
+--   vim.api.nvim_call_function('remote#host#RegisterPlugin', {
+--     'node',
+--     path .. '/rplugin/node/neovim-registry',
+--     {},
+--   })
+-- end
+
+local function register_functions()
+  local path = require('cursorless.utils').cursorless_nvim_path()
+  -- revert to using forward slashes as works when passed to remote#host#RegisterPlugin()
+  if require('cursorless.utils').is_win() then
+    path = path:gsub('\\', '/')
+  end
+  vim.api.nvim_call_function('RegisterFunctions', { path })
+end
+
+-- this triggers loading the node process as well as calling one function
+-- in the cursorless-neovim, command-server and neovim-registry extensions
+-- in order to initialize them
+local function load_extensions()
+  vim.api.nvim_call_function('CursorlessLoadExtension', {})
+  vim.api.nvim_call_function('CommandServerLoadExtension', {})
+
+  if os.getenv('CURSORLESS_MODE') == 'test' then
+    -- make sure cursorless is loaded before starting the tests
+    -- see https://neovim.io/doc/user/various.html#%3Asleep
+    vim.cmd([[sleep 5]])
```

Comment:
do we really need such a long sleep? is there no way harness can wait for cursorless to be loaded? Ie by looking in the registry? We could allow the registry to return a promise that resolves when the given extension is registered?

Reply:
We could change the loading of the extension from async to sync, but then it means it would slow down the loading of neovim in a non debug scenario. Note that this would be at the lua level. 

Also note that this timeout is only for running the tests, not for actually debugging the extension without running the tests.